pkg/dynamicinstrumentation/codegen/templates.go

```diff
@@ -0,0 +1,202 @@
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+
+package codegen
+
+var forcedVerifierErrorTemplate = `
+int illegalDereference = *(*(*ctx->regs[0]));
+`
+
+var headerTemplateText = `
+// Name={{.Name}} ID={{.ID}} TotalSize={{.TotalSize}} Kind={{.Kind}}
+// Write the kind and size to output buffer
+param_type = {{.Kind}};
+bpf_probe_read_kernel(&event->output[outputOffset], 1, &param_type);
+param_size = {{.TotalSize}};
+bpf_probe_read_kernel(&event->output[outputOffset+1], 2, &param_size);
+outputOffset += 3;
+`
+
+// The length of slices aren't known until parsing, so they require
+// special headers to read in the length dynamically
+var sliceRegisterHeaderTemplateText = `
+// Name={{.Parameter.Name}} ID={{.Parameter.ID}} TotalSize={{.Parameter.TotalSize}} Kind={{.Parameter.Kind}}
+// Write the slice kind to output buffer
+param_type = {{.Parameter.Kind}};
+bpf_probe_read_kernel(&event->output[outputOffset], 1, &param_type);
+// Read slice length and write it to output buffer
+bpf_probe_read_user(&param_size, 8, &ctx->regs[{{.Parameter.Location.Register}}+1]);
+bpf_probe_read_kernel(&event->output[outputOffset+1], 2, &param_size);
+outputOffset += 3;
+
+slice_length = param_size;
+
+for (i = 0; i < MAX_SLICE_LENGTH; i++) {
+    if (i >= slice_length) {
+        break;
+    }
+    {{.SliceTypeHeaderText}}
+}
+`
+
+// The length of slices aren't known until parsing, so they require
+// special headers to read in the length dynamically
+var sliceStackHeaderTemplateText = `
+// Name={{.Parameter.Name}} ID={{.Parameter.ID}} TotalSize={{.Parameter.TotalSize}} Kind={{.Parameter.Kind}}
+// Write the slice kind to output buffer
+param_type = {{.Parameter.Kind}};
+bpf_probe_read_kernel(&event->output[outputOffset], 1, &param_type);
+// Read slice length and write it to output buffer
+bpf_probe_read_user(&param_size, 8, &ctx->regs[29]+{{.Parameter.Location.StackOffset}}+8]);
+bpf_probe_read_kernel(&event->output[outputOffset+1], 2, &param_size);
+outputOffset += 3;
+
+slice_length = param_size;
+if (slice_length > MAX_SLICE_LENGTH) {
+    slice_length = MAX_SLICE_LENGTH;
+}
+
+for (i = 0; i < slice_length; i++) {
+    {{.SliceTypeHeaderText}}
+}
+`
+
+// The length of strings aren't known until parsing, so they require
+// special headers to read in the length dynamically
+var stringRegisterHeaderTemplateText = `
+// Name={{.Name}} ID={{.ID}} TotalSize={{.TotalSize}} Kind={{.Kind}}
+// Write the string kind to output buffer
+param_type = {{.Kind}};
+bpf_probe_read_kernel(&event->output[outputOffset], 1, &param_type);
+
+// Read string length and write it to output buffer
+bpf_probe_read_user(&param_size, 8, &ctx->regs[{{.Location.Register}}+1]);
+
+// Limit string length
+__u16 string_size_{{.ID}} = param_size;
+if (string_size_{{.ID}} > MAX_STRING_SIZE) {
+    string_size_{{.ID}} = MAX_STRING_SIZE;
+}
```

Comment:
If you set MAX_STRING_SIZE to a power of two, then `string_size & (MAX_STRING_SIZE - 1)` should constrain the value.
Branches increase verifier complexity, which may become an issue if these programs get more complex.